### ISSUE TITLE
Kavorka requires Perl >= 5.014

### ISSUE THREAD

**Issue by srchulo** [lukec/stripe-perl#28]
I can't install Net::Stripe on any of my machines because Kavorka requires Perl >= 5.014. Since you're just using it for method signatures, is there any other method signatures model that would suffice?

**Comment by resynthesize** [comment]
Just a +1 that my company also cannot use Net::Stripe due to the kavorka 5.14 dependency.  We're on CentOS 6 w/ 5.10.  Offering a fallback option for older perls would be great.

**Comment by daveaiello** [comment]
I agree with resynthesize.  We had a site set up that used Net::Stripte before the Kavorka dependency crept in.  We can't rebuild the entire site on a new version of CentOS just to get to a version of Perl that's compatible with Kavorka.

**Comment by NacreData** [comment]
I am successfully using Stripe::Perl on several systems with old Perl using http://perlbrew.pl 

It looks to me like removing this dependency would be a significant fork of the project. It's on my opensource/volunteer to-do list. If i get around to it, I would probably strip out a ton of other dependencies which are not strictly necessary. But things can linger on my unpaid to-do list for a rather long time, so I'd certainly encourage using a non-system Perl or devoting resources from your company to begin such a fork if needed.

**Comment by dmaglinte** [comment]
I'm using perlbrew perl-5.14.4. I was able to install perl, Net::Stripe and all dependencies fine but I get this error when running the stripe test (live.t):

Use of uninitialized value $e in die at ../lib/Net/Stripe.pm line 1541.
malformed JSON string, neither array, object, number, string or atom, at character offset 0 (before "LWP will support htt...") at ../lib/Net/Stripe.pm line 1529.
        ...propagated at ../lib/Net/Stripe.pm line 1541.

When I run the same code on my local machine using active perl, I get no errors. Any ideas? When I comment out the affected lines, It gets in further JSON errors.

**Comment by NacreData** [comment]
That sounds like the when you are missing some of the SSL/TOS dependencies. There is one which does not cleanly the install, I noticed just the other day. I have not had enough chance to really documented well but my searching around seemed to indicate that it had to do with a test suite which was failing because of missing certificates or certificates which had the wrong bit length, and I was able to do a force install on the needed modules to make it work

devin
--
contact info: http://nacredata.com/devin

**Comment by dmaglinte** [comment]
Thanks Devin. Do you recommend that I just re-install Net::Stripe and dependencies and force the output a file to be able to identify the culprit?

**Comment by NacreData** [comment]
That could work if you could search through the output for SSL crypt or whatever it is that is not installing properly


devin
--
contact info: http://nacredata.com/devin

**Comment by dmaglinte** [comment]
For others on this thread running into issues installing Net::Stripe and opting to use perlbrew to install an older version of perl > 5.14 (to conform to Kavorka prerequisites), cpan does not install a dependency. Installing LWP::Protocol::https should fix the problem. This is a dependency of LWP::Request. Running lwp-request https://google.com/ in the command line should confirm this.

**Comment by youradds** [comment]
This is bonkers. I've just setup a who new server, just so I can use this module (as I was only on 5.10, which wasn't new enough for Kavorka. Anyway, EVERYTHING else works, apart from this module! I can't install Kavorka , as I get an error when compiling:

> Running make test
> PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
> t/01basic.t .................... ok
> t/02named-functions.t .......... Can't locate object method "_set_declared_name" via package "Kavorka::Sub::Fun" at /root/.cpan/build/Kavorka-0.037-ICCO4K/blib/lib/Kavorka/Sub.pm line 182.
> t/02named-functions.t .......... Dubious, test returned 255 (wstat 65280, 0xff00)
> No subtests run
> t/03anon-functions.t ........... Can't locate object method "declared_name" via package "Kavorka::Sub::Fun" at /root/.cpan/build/Kavorka-0.037-ICCO4K/blib/lib/Kavorka/Sub.pm line 48.
> t/03anon-functions.t ........... Dubious, test returned 255 (wstat 65280, 0xff00)


etc etc. After some digging, it seems like there is a bug with another module that Kavorka requires, that doesn't work with > 5.21.x . PLEASE tell me someone has created a version of Net::Stripe that doesn't require that module? This is doing my nut in!

Thanks

Andy

**Comment by srchulo** [comment]
Hey Andy,

I don't think there is another module. I don't know if this is an option for you or not, but I used perlbrew on my server and that allowed me to use this module with perl 5.22 and no issues.

**Comment by youradds** [comment]
Thanks. I ended up getting some help on PerlMonks with regards to patching **Devel::CallParser**, which  in turn was stopping **Kavorka** from installing, and in turn stopping **Net::Stripe** from being installed. Here is the link, in case anyone else needs a pointer in the future:

http://perlmonks.org/?node_id=1186217

**Comment by strangelittlemonkey** [comment]
I'm getting hit by this trying to install with Perl 5.24.3

**Comment by yahermann** [comment]
I was just reviewing the source code, with an eye toward removing the Kavorka dependency.

Wouldn't this, for example:

`
    method post_charge(Int :$amount,
                       Str :$currency,
                       Net::Stripe::Customer|HashRef|Str :$customer?,
                       Net::Stripe::Card|Net::Stripe::Token|Str|HashRef :$card?,
                       Str :$description?,
                       HashRef :$metadata?,
                       Bool :$capture?,
                       Str :$statement_description?,
                       Int :$application_fee?
                     ) {
        my $charge = Net::Stripe::Charge->new(amount => $amount,
                                              currency => $currency,
                                              customer => $customer,
                                              card => $card,
                                              description => $description,
                                              metadata => $metadata,
                                              capture => $capture,
                                              statement_description => $statement_description,
                                              application_fee => $application_fee
                                          );
        return $self->_post('charges', $charge);
     }
`

Simply be a matter of changing to:

`
    sub post_charge {
       my ($self, $amount, $currency, $customer, $card, $description, $metadata, $capture, $statement_description, $application_fee) = @_;
       my $charge = Net::Stripe::Charge->new(amount => $amount,
                                              currency => $currency,
                                              customer => $customer,
                                              card => $card,
                                              description => $description,
                                              metadata => $metadata,
                                              capture => $capture,
                                              statement_description => $statement_description,
                                              application_fee => $application_fee
                                          );
         return $self->_post('charges', $charge);
    }
`

and then leaning on Moose's type constraints?

**Comment by sherrardb** [comment]
@yahermann, i may be missing something obvious here, but within the provided snippet, how would one "convert" the current Kavorka signatures into Moose type constraints?

**Comment by sherrardb** [comment]
@yahermann, indeed i was missing something obvious. after getting back to doing some coding, i see that the Moose type constraints are already in place, essentially mirroring the Kavorka signatures. 

but i believe that your suggested approach still falls victim to the problems i highlighted in #80, namely that using a role-based architecture means that you cannot reliably pull in 'has' constraints without also adding a significant abstraction layer for object creation. please have a look at my comments from late October and early November 2017.

**Comment by sherrardb** [comment]
FTR, you have my vote as well for removing the Kavorka dependency, but i would like to point out that there are other existing contexts which rely on the Kavorka validation, which would have to be addressed.

- arguments that are used locally in the method, but not passed along to the object constructor, such as $customer here:

```
    method post_customer(Net::Stripe::Customer|Str :$customer?,
                         Int :$account_balance?,
                         Net::Stripe::Card|Net::Stripe::Token|Str|HashRef :$card?,
                         Str :$coupon?,
                         Str :$default_card?,
                         Str :$description?,
                         Str :$email?,
                         HashRef :$metadata?,
                         Str :$plan?,
                         Int :$quantity?,
                         Int|Str :$trial_end?) {
...
        $customer = Net::Stripe::Customer->new(account_balance => $account_balance,
                                               card => $card,
                                               coupon => $coupon,
                                               description => $description,
                                               email => $email,
                                               metadata => $metadata,
                                               plan => $plan,
                                               quantity => $quantity,
                                               trial_end => $trial_end);
```
- methods that are not related to an object constructor
```
    method delete_invoiceitem(Net::Stripe::Invoiceitem|Str :$invoice_item) {
    method get_upcominginvoice(Net::Stripe::Customer|Str $customer) {
```
etc.

**Comment by inurear** [comment]
Hello Everyone... This worked for me - but i'm not advanced enough to explain specifically why:

`cpanm Alt::Devel::CallParser::ButWorking`

This installs a module that provides a work-around patch for a bug in Kavorka that the original author isn't patching. If you install that, then reinstall Kavorka or Net::Stripe you should be in the clear.

Details here (last post has the patch): [https://rt.cpan.org/Public/Bug/Display.html?id=110623](https://rt.cpan.org/Public/Bug/Display.html?id=110623)

**Comment by inurear** [comment]
Hello Everyone... This worked for installing Net::Stripe - but it appears to cause issues in using Stripe and I can't explain why:

`cpanm Alt::Devel::CallParser::ButWorking`

This installs a module that provides a work-around patch for a bug in Kavorka that the original author isn't patching. If you install that, then reinstall Kavorka or Net::Stripe you should be in the clear.

Details here (last post has the patch): [https://rt.cpan.org/Public/Bug/Display.html?id=110623](https://rt.cpan.org/Public/Bug/Display.html?id=110623)

However it should be noted that an undescribed error comes back when sending request to Stripe as a result:

` at /usr/local/share/perl/5.26.1/Net/Stripe.pm line 698.`
